blockchain/chain_sync/src/sync.rs

```diff
@@ -136,306 +128,226 @@ where
+        cs: Arc<ChainStore<DB>>,
+        network: SyncNetworkContext,
+        peer_id: PeerId,
+        tsk: TipsetKeys,
+        channel: Sender<(PeerId, FullTipset)>,
+    ) {
+        match Self::fetch_full_tipset(cs.as_ref(), &network, peer_id.clone(), &tsk).await {
+            Ok(fts) => {
+                channel.send((peer_id, fts)).await;
+            }
+            Err(e) => {
+                debug!("Failed to fetch full tipset from peer ({}): {}", peer_id, e);
+            }
+        }
+    }
+
+    /// Spawns a new sync worker and pushes the state to the `ChainSyncer`
+    async fn spawn_worker(&mut self, channel: Receiver<Arc<Tipset>>) -> JoinHandle<()> {
+        let state = Arc::new(RwLock::new(SyncState::default()));
+
+        // push state to managed states in Syncer.
+        self.worker_state.write().await.push(state.clone());
+        SyncWorker {
+            state,
+            beacon: self.beacon.clone(),
+            state_manager: self.state_manager.clone(),
+            chain_store: self.chain_store.clone(),
+            network: self.network.clone(),
+            genesis: self.genesis.clone(),
+            bad_blocks: self.bad_blocks.clone(),
+            peer_manager: self.peer_manager.clone(),
+        }
+        .spawn(channel)
+        .await
     }
 
     /// informs the syncer about a new potential tipset
     /// This should be called when connecting to new peers, and additionally
     /// when receiving new blocks from the network
-    pub async fn inform_new_head(&mut self, peer: PeerId, fts: &FullTipset) -> Result<(), Error> {
+    pub async fn inform_new_head(&mut self, peer: PeerId, ts: &FullTipset) -> Result<(), Error> {
         // check if full block is nil and if so return error
-        if fts.blocks().is_empty() {
+        if ts.blocks().is_empty() {
             return Err(Error::NoBlocks);
         }
         // TODO: Check if tipset has height that is too far ahead to be possible
 
-        for block in fts.blocks() {
+        for block in ts.blocks() {
             if let Some(bad) = self.bad_blocks.peek(block.cid()).await {
                 warn!("Bad block detected, cid: {:?}", bad);
                 return Err(Error::Other("Block marked as bad".to_string()));
             }
-            // validate message data
-            self.validate_msg_meta(block)?;
         }
-        // TODO: Publish LocalIncoming blocks
 
         // compare target_weight to heaviest weight stored; ignore otherwise
-        let best_weight = match self.chain_store.heaviest_tipset() {
-            Some(ts) => ts.weight().clone(),
-            None => Zero::zero(),
-        };
-        let target_weight = fts.weight();
-
-        if target_weight.gt(&best_weight) {
-            self.set_peer_head(peer, Arc::new(fts.to_tipset())).await?;
+        let candidate_ts = self
+            .chain_store
+            .heaviest_tipset()
+            .await
+            .map(|heaviest| ts.weight() >= heaviest.weight())
```

Comment:
Shouldn't this just be `ts.weight() > heaviest.weight()`? Do we want to validate and set peer head when weight is equal?

Reply:
Nah, intentional that if the weight is the same it will also be considered. Also the logic matches Lotus (Although they setup the function differently)